File: oeis-research/Amiram Eldar/Let k be the n-th perfect number, a(n) is the least number m such that k*d*m + 1 is prime for all of the proper divisors d of k so their product is a Carmichael number/lucas-carmichael_sieve.jl
#!/usr/bin/julia

# Let k = A000396(n) be the n-th perfect number, a(n) is the least number m such that k*d*m - 1 is prime for all of the proper divisors d of k so their product is a Lucas-Carmichael number.

# Lucas-Carmichael version of:
#   https://oeis.org/A319008

# Known terms:
#   1, 219, 10405375365

# Inspired by the PARI program by David A. Corneth from OEIS A372238.

# See also:
#   https://oeis.org/A372238/a372238.gp.txt

using Primes

const PERFECT_N = 496
const DIVISORS = [1, 2, 4, 8, 16, 31, 62, 124, 248]

function isrem(m::UInt128, p, n)
    return all(d -> (PERFECT_N*m*d-1) % p != 0, DIVISORS)
end

function remaindersmodp(p, n)
    filter(m -> isrem(m |> UInt128, p, n), 0:p-1)
end

function mulmod(a::UInt128, b::UInt128, m::Integer)

    # https://discourse.julialang.org/t/modular-multiplication-without-overflow/90421/4

    magic1 = (UInt128(0xFFFFFFFF) << 32)
    magic2 = UInt128(0x8000000000000000)

    if iszero(((a | b) & magic1))
        return (a * b) % m
    end

    d = zero(UInt128)
    mp2 = m >> 1

    if a >= m; a %= m; end
    if b >= m; b %= m; end

    for _ in 1:64
        (d > mp2) ? d = ((d << 1) - m) : d = (d << 1)
        if !iszero(a & magic2)
            d += b
        end
        if d >= m
            d -= m
        end
        a <<= 1
    end

    return d
end

function mulmod(a::Integer, b::Integer, m::Integer)
    sa, sb = UInt128.(unsigned.((a,b)))
    return mulmod(sa, sb, m)
end

function chinese(a1, m1, a2, m2)
    M = lcm(m1, m2)
    return [
         (mulmod(mulmod(a1, invmod(div(M, m1), m1), M), div(M, m1), M) +
          mulmod(mulmod(a2, invmod(div(M, m2), m2), M), div(M, m2), M)) % M, M
    ]
end

function remainders_for_primes(n, primes)

    res = [[0, 1]]

    for p in primes

        rems = remaindersmodp(p, n)

        if (length(rems) == 0)
            rems = [0]
        end

        nres = []
        for r in res
            a = r[1]
            m = r[2]
            for rem in rems
                push!(nres, chinese(a, m, rem, p))
            end
        end
        res = nres
    end

    res = map(x -> x[1], res)
    sort!(res)
    return res
end

function deltas(arr)
    prev = 0
    D = []
    for k in arr
        push!(D, k - prev)
        prev = k
    end
    return D
end

function generate(n)

    maxp = 11

    n >= 3 && (maxp = 19)
    n >= 4 && (maxp = 29)
    n >= 5 && (maxp = 37)

    P = primes(maxp)

    R = remainders_for_primes(n, P)
    D = deltas(R)
    s = prod(P)

    while (D[1] == 0)
        popfirst!(D)
    end

    push!(D, R[1] + s - R[length(R)])

    m      = R[1] |> UInt128
    D_len  = length(D)

    for j in 0:10^18

        if (all(d -> isprime(PERFECT_N*m*d-1), DIVISORS))
            return m |> Int128
        end

        if (j % 10^7 == 0 && j > 0)
            println("[$j] Searching for a($n) with m = $m")
        end

        m += D[(j % D_len) + 1]
    end
end

const n = 3
println(generate(n))

File: oeis-research/Amiram Eldar/Let k be the n-th perfect number, a(n) is the least number m such that k*d*m + 1 is prime for all of the proper divisors d of k so their product is a Carmichael number/lucas-carmichael_sieve.pl
#!/usr/bin/perl

# Let k = A000396(n) be the n-th perfect number, a(n) is the least number m such that k*d*m - 1 is prime for all of the proper divisors d of k so their product is a Lucas-Carmichael number.

# Lucas-Carmichael version of:
#   https://oeis.org/A319008

# Known terms:
#   1, 219, 10405375365

# Inspired by the PARI program by David A. Corneth from OEIS A372238.

# See also:
#   https://oeis.org/A372238/a372238.gp.txt

# Lower-bounds:
#   a(4) > 1116642267975

use 5.036;
use ntheory     qw(:all);
use Time::HiRes qw (time);
use List::Util  qw(all any);

my $PERFECT_N;
my @DIVISORS;

sub isrem($m, $p, $n) {
    vecall { (vecprod($PERFECT_N, $m, $_) - 1) % $p != 0 } @DIVISORS;
}

sub remaindersmodp($p, $n) {
    grep { isrem($_, $p, $n) } (0 .. $p - 1);
}

sub remainders_for_primes($n, $primes) {

    my $res = [[0, 1]];

    foreach my $p (@$primes) {

        my @rems = remaindersmodp($p, $n);

        my @nres;
        foreach my $r (@$res) {
            foreach my $rem (@rems) {
                push @nres, [chinese($r, [$rem, $p]), lcm($p, $r->[1])];
            }
        }
        $res = \@nres;
    }

    sort { $a <=> $b } map { $_->[0] } @$res;
}

sub deltas ($integers) {

    my @deltas;
    my $prev = 0;

    foreach my $n (@$integers) {
        push @deltas, $n - $prev;
        $prev = $n;
    }

    return \@deltas;
}

sub generate($n) {

    my $maxp = 11;

    $maxp = 23 if ($n >= 3);
    $maxp = 29 if ($n >= 4);

    my @primes = @{primes($maxp)};

    my @r = remainders_for_primes($n, \@primes);
    my @d = @{deltas(\@r)};
    my $s = vecprod(@primes);

    while ($d[0] == 0) {
        shift @d;
    }

    push @d, $r[0] + $s - $r[-1];

    my $m      = $r[0];
    my $d_len  = scalar(@d);
    my $t0     = time;
    my $prev_m = $m;

    for (my $j = 0 ; ; ++$j) {

        if (vecall { is_prime(vecprod($m, $PERFECT_N, $_) - 1) } @DIVISORS) {
            return $m;
        }

        if ($j % 1e7 == 0 and $j > 0) {
            my $tdelta = time - $t0;
            say "[$j] Searching for a($n) with m = $m";
            say "Performance: ", (($m - $prev_m) / 1e9) / $tdelta, " * 10^9 terms per second";
            $t0     = time;
            $prev_m = $m;
        }

        $m += $d[$j % $d_len];
    }
}

my @perfect_numbers = (6, 28, 496, 8128, 33550336, 8589869056, 137438691328, 2305843008139952128);

foreach my $n (1 .. $#perfect_numbers) {

    $PERFECT_N = $perfect_numbers[$n - 1];
    @DIVISORS  = grep { $_ < $PERFECT_N } divisors($PERFECT_N);

    say "a($n) = ", generate($n);
}

__END__
a(1) = 4
a(2) = 219
a(3) = 10405375365
Searching for a(4) with m = 32842303065
Performance: 1.5614861010746 * 10^9 terms per second
Searching for a(4) with m = 65684651175
Performance: 1.56154555876966 * 10^9 terms per second
Searching for a(4) with m = 98527241490
Performance: 1.5464539801719 * 10^9 terms per second
